### DISCUSSION TITLE
Clamping CIELAB to the visible spectrum?

### DISCUSSION THREAD

**Discussion by blue-j** [colour-science/colour#1141] [Q&A] [answered]
I wrote a def that does this by roundtripping through sd with the visible range of wavelengths:

```
def clamp_cielab_to_visible_spectrum(cielab, illuminant):

    illuminant_XY = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer'][illuminant]
    # convert CIELAB to XYZ
    xyz = colour.Lab_to_XYZ(cielab, illuminant_XY)

    # convert XYZ to spectral distribution
    sd = colour.XYZ_to_sd(xyz, method='Jakob 2019')

    # clamp to the spectral distribution of the visible range (360 nm to 830 nm)
    # values for visible range from CIE Colorimetry (4th edition) CIE 015:2018 page 21
    clamped_sd = colour.SpectralDistribution({wavelength: sd[wavelength] for wavelength in sd.wavelengths if 360 <= wavelength <= 830})

    # convert clamped spectral distribution back to XYZ
    xyz_clamped = colour.sd_to_XYZ(clamped_sd)

    # convert clamped XYZ back to CIELAB
    cielab = colour.XYZ_to_Lab(xyz_clamped, illuminant_XY)

    return cielab
```

Foolish?  Inefficient?  I can tell you now this takes like 20 seconds, so I am hoping there is a better way.

Why are you doing this, J?  Because I am actually honoring the formal spec of CIELAB and allowing a* and b* to be input unbounded, but offer the option to push them back into the visible.

I'm eager for your thoughts!

- J

**Comment by tjdcs** [comment]
Actually, there's nothing improper about color stimuli that have a greater chroma value than the spectrum locus might imply. It's been seen in research! And is relatively easy to produce under the right conditions on an HDR display. 

http://markfairchild.org/PDFs/PAP22.pdf

**Reply by blue-j** [reply]
Thanks so much for sharing.  I appreciate it!  I am wondering, my clamping uses wavelengths of the visible spectrum, not the spectral locus proper.  Does this have any implications?  I am unsure after reviewing the paper, which doesn't mention wavelengths.  - J

**Reply by thomasmansencal** [reply, marked as answer]
We have the `colour.is_within_visible_spectrum` definition which is relevant: https://github.com/colour-science/colour/blob/develop/colour/volume/spectrum.py#L381

The hull computation is costly but only on first invocation, it is cached after that.
